### PR TITLE
DTPPSDK-239: Add new <BraintreePayPalButtons /> component

### PR DESCRIPTION
**Pull request by borodovisin**
This PR adds a new `<BraintreePayPalButtons />` component for Braintree merchants who would like to render the PayPal Buttons using the JS SDK (https://developer.paypal.com/braintree/docs/guides/paypal/client-side).

### Code Example

There's a couple things to highlight with the following code example:
- Braintree SDK integrations require both server-side and client-side code. A new Braintree auth token is required for each page render. Similar to hosted-fields, we use the `data-client-token` attribute for accepting this value. This can be server-side rendered directly into the JSX like in the code snippet below. Or it can be added at a later time using the dispatcher feature of the provider.
- The `paypalCheckoutInstance` needs to be made available to the end user. This solution is decorating the `actions` for createOrder and onApprove with this instance as `actions.braintree` (ex: `actions.braintree.createPayment()`).

```jsx
<PayPalScriptProvider
    options={{
        "client-id": "test",
        "data-client-token": "abc123xyz==",
    }}
>
    <BraintreePayPalButtons
        createOrder={(data, actions) => {
            return actions.braintree.createPayment({
                flow: "checkout",
                amount: "10.0",
                currency: "USD",
                intent: "capture"
            });
        }}
        onApprove={(data, actions) => {
            return actions.braintree.tokenizePayment(data)
                .then((payload) => {
                    // call server-side endpoint to finish the sale
                })
        }
    />
</PayPalScriptProvider>
```

### Storybook Screenshot

storybook-screenshot

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -1,2 +1,20 @@
 // Common reference to the script identifier
 export const SCRIPT_ID = "data-react-paypal-script-id";
+export const DATA_CLIENT_TOKEN = "data-client-token";
+
+export const EMPTY_PROVIDER_CONTEXT_ERROR_MESSAGE =
+    "usePayPalScriptReducer must be used within a PayPalScriptProvider";
+export const EMPTY_PROVIDER_CONTEXT_CLIENT_TOKEN_ERROR_MESSAGE =
+    "A client token wasn't found in the provider parent component";
+
+export const BRAINTREE_CLIENT_SCRIPT_ID = "braintree-client-script";
```
Comment (gregjopa) [inline src/constants.ts:10]: Let's remove this script id values. We don't need them for anything specific and I would like to keep the bundle size as small as possible.
```suggestion
```

File: src/constants.ts [hunk 2]
```diff
@@ -1,2 +1,20 @@
...
+export const BRAINTREE_PAYPAL_CHECKOUT_SCRIPT_ID =
```
Comment (gregjopa) [inline src/constants.ts:11]:
```suggestion
```

File: src/types/braintreePayPalButtonTypes.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import type { Client, PayPalCheckout } from "braintree-web";
+import type { PayPalButtonsComponentProps } from "./paypalButtonTypes";
+import { UnknownObject } from "@paypal/paypal-js/types/apis/subscriptions";
```
Comment (gregjopa) [inline src/types/braintreePayPalButtonTypes.ts:3]: Let's avoid importing this. It's just using `Record<string, unknown>`. Let's use that instead
```suggestion
```

File: src/utils.ts [hunk 4]
```diff
@@ -10,11 +21,29 @@ export function getPayPalWindowNamespace(
...
+/**
+ * Creates a numeric hash based on the string input
+ *
+ * @param str the source string to generate the hash
+ * @returns a hash numeric ode
  */
 export function hashStr(str: string): number {
     let hash = 0;
-    for (let i = 0; i < str.length; i++) {
+    for (let i = 0; i < (str || "").length; i++) {
```
Comment (gregjopa) [inline src/utils.ts:46]: Please revert. The string argument is required. 

```suggestion
    for (let i = 0; i < (str).length; i++) {
```

File: src/utils.test.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import {
+    getPayPalWindowNamespace,
+    getBraintreeWindowNamespace,
+    hashStr,
+} from "./utils";
+
+describe("testing getPayPalWindowNamespace util function", () => {
```
Comment (gregjopa) [inline src/utils.test.js:7]: Let's follow the pattern of using the function name as the describe block value.

```suggestion
describe("getPayPalWindowNamespace", () => {
```

File: src/utils.test.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    const paypal = {
+        Buttons: jest.fn(),
+    };
+
+    beforeAll(() => {
+        window.paypal = paypal;
+    });
+
+    test("should return the paypal namespace", () => {
+        expect(getPayPalWindowNamespace("paypal")).toEqual(paypal);
```
Comment (gregjopa) [inline src/utils.test.js:17]: I think you could use `window.paypal` directly and eliminate the need for the extra variable.

```js
    beforeAll(() => {
        window.paypal = {
            Buttons: jest.fn(),
        };
    });
```

```suggestion
        expect(getPayPalWindowNamespace("paypal")).toEqual(window.paypal);
```

File: src/utils.test.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    });
+
+    test("should not found the namespace", () => {
+        expect(getBraintreeWindowNamespace("testNamespace")).toBeUndefined();
+    });
+});
+
+describe("testing hashStr util function", () => {
+    test("should match the hash from the argument string", () => {
+        expect(hashStr("react")).toMatchSnapshot();
```
Comment (gregjopa) [inline src/utils.test.js:46]: Thanks for adding tests for the hashStr function! Can you add an assertion for a real world value? I believe we pass the `JSON.stringify(options)` from the `loadScript` function. Let's add an assertion that tests it.

File: src/utils.test.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    test("should return the paypal namespace", () => {
+        expect(getPayPalWindowNamespace("paypal")).toEqual(paypal);
+    });
+
+    test("should not found the namespace", () => {
+        expect(getPayPalWindowNamespace("testNamespace")).toBeUndefined();
+    });
+});
+
+describe("testing getBraintreeWindowNamespace util function", () => {
```
Comment (gregjopa) [inline src/utils.test.js:25]:
```suggestion
describe("getBraintreeWindowNamespace", () => {
```

File: src/utils.test.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    test("should return the paypal namespace", () => {
+        expect(getBraintreeWindowNamespace("braintree")).toEqual(braintree);
+    });
+
+    test("should not found the namespace", () => {
+        expect(getBraintreeWindowNamespace("testNamespace")).toBeUndefined();
+    });
+});
+
+describe("testing hashStr util function", () => {
```
Comment (gregjopa) [inline src/utils.test.js:44]:
```suggestion
describe("hashStr", () => {
```

File: package.json [hunk 10]
```diff
@@ -63,6 +63,7 @@
         "@storybook/react": "^6.3.6",
         "@storybook/storybook-deployer": "^2.8.10",
         "@testing-library/react": "^12.0.0",
+        "@types/braintree-web": "^3.75.6",
```
Comment (gregjopa) [inline package.json:66]: I would like us to investigate adding these braintree-web types ourselves to this library. That way we have control over them. It looks like we are only depending on these imports:

```
import type { Client, PayPalCheckout } from "braintree-web";
```

File: src/components/BraintreePayPalButton.tsx [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+/**
+ * React functional component to create a PayPal button using the Braintree
+ * This component is a wrapper of the PayPalButtons and override the logic
+ * of the createOrder and onApprove callback function from the SDK
+ *
+ * @param param0 the props of the component
+ * @returns an JSX code to translate to the DOM
+ */
+export const BraintreePayPalButton: FC<BraintreePayPalButtonsComponentProps> =
```
Comment (gregjopa) [inline src/components/BraintreePayPalButton.tsx:26]: I was thinking more about this and let's use the plural version to match `<PayPalButtons />`. It is possible to use a "smart" integration with braintree where there will be more than one button.

```suggestion
export const BraintreePayPalButtons: FC<BraintreePayPalButtonsComponentProps> =
```

File: src/components/BraintreePayPalButton.tsx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                const clientToken = providerContext?.options[
+                    DATA_CLIENT_TOKEN
+                ] as string;
+
+                const braintreeNamespace = getBraintreeWindowNamespace();
+                const clientInstance = await braintreeNamespace.client.create({
+                    authorization: clientToken,
+                });
+
+                setPayPalCheckoutInstance(
```
Comment (westeezy) [inline src/components/BraintreePayPalButton.tsx:58]: Would this make more sense to store in the provider? That would also allow consumers access to the braintree client as an escape hatch if needed.

File: src/components/BraintreePayPalButton.tsx [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+            });
+        }, [providerContext?.options]);
+
+        /**
+         * Override the createOrder callback to send the PayPal checkout instance as argument
+         * to the defined createOrder function for braintree component button
+         *
+         * @param braintreeButtonProps the component button options
+         */
+        const decorateCreateOrder = (
```
Comment (westeezy) [inline src/components/BraintreePayPalButton.tsx:72]: Would these decorators make sense as utility functions? I think moving it out might help improve readability as well to keep the component a bit more focused.

File: src/components/PayPalButtons.tsx [hunk 14]
```diff
@@ -149,6 +123,7 @@ export const PayPalButtons: FunctionComponent<PayPalButtonsComponentProps> = ({
         });
 
         return closeButtonsComponent;
+        // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (westeezy) [inline src/components/PayPalButtons.tsx:126]: Was this change needed? I don't see any changes to the hook code so curious why we needed to disable the exhaustive deps on this PR.

Comment (westeezy) [reply src/components/PayPalButtons.tsx:126]: Guess I'm seeing it in a few spots possibly related to moving a hook. Did a previous unit of work add this rule but not update untouched files?

Comment (gregjopa) [reply src/components/PayPalButtons.tsx:126]: > Did a previous unit of work add this rule but not update untouched files?

PR #138 added the new react hooks eslint plugin. The main branch currently has eslint warnings related to this and we forgot to add these before. It's basically keeping eslint quiet since we are purposely leaving props out with the `forceReRender` strategy.

File: src/components/BraintreePayPalButton.tsx (hunk 12 above)
Comment (gregjopa) [reply src/components/BraintreePayPalButton.tsx:58]: Great recommendation @westeezy!

File: package.json (hunk 10 above)
Comment (westeezy) [reply package.json:66]: `braintree-web` might not have typescript typings in the repo just yet. I might be wrong though just trying to recall if its been added or still on the roadmap to start soon

File: src/components/PayPalButtons.tsx (hunk 14 above)
Comment (borodovisin) [reply src/components/PayPalButtons.tsx:126]: I had to use that comment to avoid the warning cuz eslint-plugin-react-hooks, Greg mention we don't need all the options as dependencies inside the useEffect, I think we change the dependencies inside a PR. Not completely sure, If I remove the comment I'll see eslint warning with missing dependencies. Any thought or suggestions are welcome to avoid this.

Comment (westeezy) [reply src/components/PayPalButtons.tsx:126]: Oh yeah thats right, I totally spaced on that. We already chatted this so I'll mark as resolved! Thanks!

File: src/components/BraintreePayPalButton.tsx (hunk 12 above)
Comment (borodovisin) [reply src/components/BraintreePayPalButton.tsx:58]: I'll need some directions here, should I put the braintree instance inside the context to allows children get access to it?
I have the same questions, not sure if that make sense or not, I guess that depends on the business and how useful will be for children components? I'm ok with either ways, only inside the state of the component or inside the context.

Comment (westeezy) [reply src/components/BraintreePayPalButton.tsx:58]: I am up for conversation around it! My suggestion was leveraging the context we setup to house the instantiated library.  My main reason to suggestion the change was more for predictability and having the instance on context as this library grows. It felt more intuitive to me personally, but absolutely open to more discussion. The use-case for end consumers to pull the instantiated client from context is more of a side-effect that might allow them some additional flexibility, but that isn't the main end goal of this change.

File: package.json (hunk 10 above)
Comment (borodovisin) [reply package.json:66]: Here what I should do?. Since we don't have types in the braintree-web project, should I keep unknown types in our project or should I create those types in our react-paypal-js project and in the second case how I guarantee the integrity of the new types from any change in the braintree-web project?

File: src/components/BraintreePayPalButton.test.js [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+            }),
+        },
+    };
+
+    loadScript.mockResolvedValue(window.paypal);
+    loadCustomScript.mockResolvedValue(window.braintree);
+};
+
+describe("Braintree PayPal button fail in mount process", () => {
+    beforeAll(setup);
```
Comment (gregjopa) [inline src/components/BraintreePayPalButton.test.js:52]: Let's use beforeEach for a clean state for each test
```suggestion
    beforeEach(setup);
```

File: src/stories/BraintreePayPalButtons.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+                            <Story />
+                        </PayPalScriptProvider>
+                    )}
+                </div>
+            );
+        },
+    ],
+};
+
+export const Default: FC = () => <BraintreePayPalButtons />;
```
Comment (gregjopa) [inline src/stories/BraintreePayPalButtons.stories.tsx:63]: Great job adding a story for this new component. I do think it would be best to show how to use `actions.braintree.createPayment()` and `actions.braintree.tokenizePayment()`. Can you update the story to show how to use those?

Here's a quick example of using `createPayment`. 
```jsx
export const Default: FC = () => (
    <BraintreePayPalButtons
        createOrder={(data, actions) => {
            return actions.braintree.createPayment({
                flow: "checkout",
                amount: 10.0,
                currency: "USD",

                intent: "capture",

                enableShippingAddress: true,
                shippingAddressEditable: false,
                shippingAddressOverride: {
                    recipientName: "Scruff McGruff",
                    line1: "1234 Main St.",
                    line2: "Unit 1",
                    city: "Chicago",
                    countryCode: "US",
                    postalCode: "60652",
                    state: "IL",
                    phone: "123.456.7890",
                },
            });
        }}
    />
);
```

Comment (gregjopa) [reply src/stories/BraintreePayPalButtons.stories.tsx:63]: I'm also getting some type errors with the above snippet:
Screen Shot 2021-08-18 at 9 48 09 AM

Comment (borodovisin) [reply src/stories/BraintreePayPalButtons.stories.tsx:63]: Done the createOrder and onApprove callbacks examples

File: src/constants.ts [hunk 17]
```diff
@@ -1,2 +1,16 @@
 // Common reference to the script identifier
 export const SCRIPT_ID = "data-react-paypal-script-id";
+export const DATA_CLIENT_TOKEN = "data-client-token";
+
+export const EMPTY_PROVIDER_CONTEXT_ERROR_MESSAGE =
+    "usePayPalScriptReducer must be used within a PayPalScriptProvider";
+export const EMPTY_PROVIDER_CONTEXT_CLIENT_TOKEN_ERROR_MESSAGE =
+    "A client token wasn't found in the provider parent component";
+
+const braintreeVersion = "3.79.1";
```
Comment (gregjopa) [inline src/constants.ts:10]: A new version was recently released so let's use the latest:

```suggestion
const braintreeVersion = "3.81.0";
```

File: src/hooks/braintreeProviderHooks.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+import { ScriptContext } from "../context/scriptProviderContext";
+import {
+    contextNotEmptyValidator,
+    contextOptionClientTokenNotEmptyValidator,
+} from "./contextValidator";
+
+/**
+ * Custom hook to get access to the ScriptProvider context
+ *
+ * @returns the state of the con text
```
Comment (gregjopa) [inline src/hooks/braintreeProviderHooks.ts:14]:
```suggestion
 * @returns the state of the context
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+            style={{ layout: "horizontal" }}
+            createOrder={(data, actions) => {
+                return actions.braintree.createPayment({
+                    flow: "checkout",
+                    amount: "10.0",
+                    currency: "USD",
+                    intent: "capture",
+                    enableShippingAddress: true,
+                    shippingAddressEditable: false,
+                    shippingAddressOverride: {
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:36]: Let's reduce this code snippet by only passing the required attrs to createPayment(). I'm thinking we can get rid of the `shippingAddressOverride` object and maybe others.

File: src/stories/BraintreePayPalButtons.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+                            line1: "1234 Main St.",
+                            line2: "Unit 1",
+                            city: "Chicago",
+                            countryCode: "US",
+                            postalCode: "60652",
+                            state: "IL",
+                            phone: "123.456.7890",
+                        },
+                    };
+                arg.createPayment({ data, actions });
```
Comment (gregjopa) [inline src/stories/BraintreePayPalButtons.stories.tsx:109]: This shouldn't be needed.
```suggestion
```

File: src/components/braintree/utils.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+ * to the defined createOrder function for braintree component button
+ *
+ * @param braintreeButtonProps the component button options
+ */
+const decorateCreateOrder = (
+    braintreeButtonProps: BraintreePayPalButtonsComponentProps,
+    payPalCheckoutInstance: BraintreePayPalCheckout
+) => {
+    if (
+        braintreeButtonProps?.createOrder != undefined &&
```
Comment (gregjopa) [inline src/components/braintree/utils.ts:16]: Let's validate that it's a function like so:
```suggestion
        typeof braintreeButtonProps.createOrder === 'function'
```

Also, can we remove the `payPalCheckoutInstance` check? We already validate it in BraintreePayPalButtons.tsx

File: src/components/braintree/utils.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+        braintreeButtonProps.onApprove = (data, actions) =>
+            braintreeOnApprove(data, {
+                ...actions,
+                braintree: payPalCheckoutInstance,
+            });
+    }
+};
+
+/**
+ * Massage needed functions to adapt to PayPal SDK types
```
Comment (gregjopa) [inline src/components/braintree/utils.ts:56]:
```suggestion
 * Use `actions.braintree` to provide an interface for the paypalCheckoutInstance through the createOrder and onApprove callbacks
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+        ...buttonProps
+    }: BraintreePayPalButtonsComponentProps) => {
+        const [providerContext, dispatch] = useBraintreeProviderContext();
+
+        useEffect(() => {
+            Promise.all([
+                loadCustomScript({ url: BRAINTREE_SOURCE }),
+                loadCustomScript({ url: BRAINTREE_PAYPAL_CHECKOUT_SOURCE }),
+            ]).then(async () => {
+                const clientToken = providerContext?.options[
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:79]:
```suggestion
                const clientToken = providerContext.options[
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+                    authorization: clientToken,
+                });
+                dispatch({
+                    type: DISPATCH_ACTION.SET_BRAINTREE_INSTANCE,
+                    value: await braintreeNamespace.paypalCheckout.create({
+                        client: clientInstance,
+                    }),
+                });
+            });
+        }, [providerContext?.options, dispatch]);
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:94]:
```suggestion
        }, [providerContext.options, dispatch]);
```

File: README.md [hunk 25]
```diff
@@ -163,7 +163,34 @@ To learn more, check out the [dynamic currency example in storybook](https://pay
...
+
+export default function App() {
+    return (
+        <PayPalScriptProvider
+            options={{
+                "client-id": "test",
+                "data-client-token": "abc123xyz==",
+            }}
+        >
+            <BraintreePayPalButtons style={{ layout: "horizontal" }} />
```
Comment (gregjopa) [inline README.md:187]:
```suggestion
            <BraintreePayPalButtons
                createOrder={(data, actions) => {
                    return actions.braintree.createPayment({
                        flow: "checkout",
                        amount: "10.0",
                        currency: "USD",
                        intent: "capture"
                    });
                }}
                onApprove={(data, actions) => {
                    return actions.braintree
                        .tokenizePayment(data)
                        .then((payload) => {
                            // call server-side endpoint to finish the sale
                        })
                }
            />
```

File: src/components/braintree/BraintreePayPalButtons.test.js [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+            render(
+                <PayPalScriptProvider>
+                    <BraintreePayPalButtons />
+                </PayPalScriptProvider>
+            );
+        } catch (ex) {
+            errorMessage = ex.message;
+        }
+        expect(errorMessage).toEqual(
+            "Cannot read property 'length' of undefined"
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.test.js:85]: I remember you bringing up this issue before. I think I understand now. We do not want this error happening. To solve, we should set `options` to an empty object by default: /Users/gjopa/projects/react-paypal-js/src/components/PayPalScriptProvider.tsx.

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+    <PayPalScriptProvider options={{ "client-id": "test" }}>
+        <BraintreePayPalButtons
+            style={{ layout: "horizontal" }}
+            createOrder={(data, actions) => {
+                return actions.braintree.createPayment({
+                    flow: "checkout",
+                    amount: "10.0",
+                    currency: "USD",
+                    intent: "capture",
+                    enableShippingAddress: true,
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:34]:
```suggestion
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+                    shippingAddressEditable: false
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:35]: Let's remove these optional attrs to shorten up the snippet.
```suggestion
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+                })
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:36]: Let's add the `onApprove` callback here as well just like the code snippet in the readme.

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+Use props for customizing your buttons. For example, here's how you would use the `style` and `createOrder` options:
+
+```jsx
+    import { PayPalScriptProvider, BraintreePayPalButtons } from "@paypal/react-paypal-js";
+
+    <PayPalScriptProvider options={{ "client-id": "test" }}>
+        <BraintreePayPalButtons
+            style={{ layout: "horizontal" }}
+            createOrder={(data, actions) => {
+                return actions.braintree.createPayment({
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:29]:
```suggestion
                // the paypalCheckoutInstance from the braintree sdk integration is added to `actions.braintree`
                return actions.braintree.createPayment({
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+                    intent: "capture",
+                    enableShippingAddress: true,
+                    shippingAddressEditable: false
+                })
+            }}
+        />;
+    </PayPalScriptProvider>
+```
+
+Notice the main difference comparing with `<PayPalButtons />` component is the [braintree paypalCheckout](https://developer.paypal.com/braintree/docs/guides/paypal/checkout-with-paypal)
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:42]: I understand the point you'd like to make here with describing `actions.braintree`. Let's explain it in the above code snippet instead using code comments.

File: src/components/braintree/utils.ts [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+ * to the defined createOrder function for braintree component button
+ *
+ * @param braintreeButtonProps the component button options
+ */
+const decorateCreateOrder = (
+    braintreeButtonProps: BraintreePayPalButtonsComponentProps,
+    payPalCheckoutInstance: BraintreePayPalCheckout
+) => {
+    if (
+        typeof braintreeButtonProps?.createOrder === "function" &&
```
Comment (gregjopa) [inline src/components/braintree/utils.ts:16]:
```suggestion
        typeof braintreeButtonProps.createOrder === "function" &&
```

File: src/components/braintree/utils.ts [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+        payPalCheckoutInstance != undefined
```
Comment (gregjopa) [inline src/components/braintree/utils.ts:17]: Let's get rid of this check.
```suggestion
```

File: src/components/braintree/utils.ts [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+ * to the defined onApprove function for braintree component button
+ *
+ * @param braintreeButtonProps the component button options
+ */
+const decorateOnApprove = (
+    braintreeButtonProps: BraintreePayPalButtonsComponentProps,
+    payPalCheckoutInstance: BraintreePayPalCheckout
+) => {
+    if (
+        braintreeButtonProps?.onApprove != undefined &&
```
Comment (gregjopa) [inline src/components/braintree/utils.ts:41]:
```suggestion
typeof braintreeButtonProps.onApprove === "function" &&
```

File: src/components/braintree/utils.ts [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+        payPalCheckoutInstance != undefined
```
Comment (gregjopa) [inline src/components/braintree/utils.ts:42]:
```suggestion
```

File: src/stories/BraintreePayPalButtons.stories.tsx [hunk 36]
```diff
@@ -0,0 +1,116 @@
...
+            ) =>
+                actions.braintree
+                    .tokenizePayment(
+                        data as BraintreePayPalCheckoutTokenizationOptions
+                    )
+                    .then((payload) => {
+                        approveSale(payload.nonce, AMOUNT).then((data) => {
+                            alert(JSON.stringify(data));
+                            // Call server-side endpoint to finish the sale
+                            return;
```
Comment (gregjopa) [inline src/stories/BraintreePayPalButtons.stories.tsx:110]:
```suggestion
```

File: src/types/braintreePayPalButtonTypes.ts [hunk 37]
```diff
@@ -0,0 +1,48 @@
...
+    payerId?: string | null;
+    paymentId?: string | null;
+    subscriptionId?: string | null;
+    authCode?: string | null;
+};
+
+export interface BraintreePayPalButtonsComponentProps
+    extends Omit<PayPalButtonsComponentProps, "createOrder" | "onApprove"> {
+    /**
+     * Override the default createOrder function from paypal-js library
```
Comment (gregjopa) [inline src/types/braintreePayPalButtonTypes.ts:30]:
```suggestion
     * The createOrder actions include the braintree sdk paypalCheckoutInstance as `actions.braintree`
```

File: src/types/braintreePayPalButtonTypes.ts [hunk 38]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    createOrder?: (
+        data: Record<string, unknown>,
+        actions: CreateBraintreeActions
+    ) => Promise<string>;
+    /**
+     * Override the default onApprove function from paypal-js library
```
Comment (gregjopa) [inline src/types/braintreePayPalButtonTypes.ts:37]:
```suggestion
     * The onApprove actions include the braintree sdk paypalCheckoutInstance as `actions.braintree`
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 39]
```diff
@@ -0,0 +1,113 @@
...
+} from "../../constants";
+import type { BraintreePayPalButtonsComponentProps } from "../../types/braintreePayPalButtonTypes";
+import { PayPalButtons } from "../PayPalButtons";
+import { useBraintreeProviderContext } from "../../hooks/braintreeProviderHooks";
+import { getBraintreeWindowNamespace } from "../../utils";
+import { DISPATCH_ACTION } from "../../types/enums";
+import { decorateActions } from "./utils";
+
+/**
+This `<BraintreePayPalButtons />` component renders the Braintree PayPal [Smart Payment Buttons](https://developer.paypal.com/docs/business/javascript-sdk/javascript-sdk-reference/#buttons).
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:17]:
```suggestion
This `<BraintreePayPalButtons />` component renders the [Braintree PayPal Buttons](https://developer.paypal.com/braintree/docs/guides/paypal/overview).
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 40]
```diff
@@ -0,0 +1,98 @@
...
+} from "../../constants";
+import type { BraintreePayPalButtonsComponentProps } from "../../types/braintreePayPalButtonTypes";
+import { PayPalButtons } from "../PayPalButtons";
+import { useBraintreeProviderContext } from "../../hooks/braintreeProviderHooks";
+import { getBraintreeWindowNamespace } from "../../utils";
+import { DISPATCH_ACTION } from "../../types/enums";
+import { decorateActions } from "./utils";
+
+/**
+This `<BraintreePayPalButtons />` component renders the [Braintree PayPal Buttons](https://developer.paypal.com/braintree/docs/guides/paypal/overview).
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:17]:
```suggestion
This `<BraintreePayPalButtons />` component renders the [Braintree PayPal Buttons](https://developer.paypal.com/braintree/docs/guides/paypal/overview) for Braintree Merchants.
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 41]
```diff
@@ -0,0 +1,98 @@
...
+import { useBraintreeProviderContext } from "../../hooks/braintreeProviderHooks";
+import { getBraintreeWindowNamespace } from "../../utils";
+import { DISPATCH_ACTION } from "../../types/enums";
+import { decorateActions } from "./utils";
+
+/**
+This `<BraintreePayPalButtons />` component renders the [Braintree PayPal Buttons](https://developer.paypal.com/braintree/docs/guides/paypal/overview) for Braintree Merchants.
+It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.
+
+Use props for customizing your buttons. For example, here's how you would use the `style` and `createOrder` options:
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:20]:
```suggestion
Use props for customizing your buttons. For example, here's how you would use the `style`, `createOrder`, and `onApprove` options:
```

File: src/components/braintree/BraintreePayPalButtons.tsx [hunk 42]
```diff
@@ -0,0 +1,98 @@
...
+                // the paypalCheckoutInstance from the braintree sdk integration is added to `actions.braintree`
+                return actions.braintree.createPayment({
+                    flow: "checkout",
+                    amount: "10.0",
+                    currency: "USD",
+                    intent: "capture"
+                })
+            }}
+            onApprove={(data, actions) => {
+                // the paypalCheckoutInstance from the braintree sdk integration is added to `actions.braintree`
```
Comment (gregjopa) [inline src/components/braintree/BraintreePayPalButtons.tsx:38]: This comment seems redundant. I think we can remove this one.
```suggestion
```

**Review by gregjopa — APPROVED** [review]
Awesome work @borodovisin! 💯

**Comment by philbirt** [comment]
Noticed this open PR yesterday and It looks like this has been approved. Any estimate on when this might get merged and cut in a new release? Thanks for building this out, its very helpful!

**Comment by gregjopa** [comment]
> Noticed this open PR yesterday and It looks like this has been approved. Any estimate on when this might get merged and cut in a new release? Thanks for building this out, its very helpful!

Thanks for the positive feedback @philbirt! We plan to publish a new version to npm tomorrow.
